### DISCUSSION TITLE
How can we automate resource migrations

### DISCUSSION THREAD

**Discussion by sinabakh** [infracost/infracost#1110] [Ideas & Feature Requests]
I'm thinking about how can we automate the process of migrating the old resources located in `internal/providers` to the new resource structure located in `internal/resources`.

**Comment by sinabakh** [comment]
The key difference I can think of is the introduction of embedded usage fields and usage schema.
To generate these data we can do static code analysis and extract such cases: `u.Get("foo_usage_key")` and do 3 steps:
1. Generate a field by converting the key string to camelCase field (but how to infer the type of the variable? Probably by looking it up in the reference usage file.)
2. Add it to the usage schema of the resource and determine the ValueType of it by looking up the reference usage file.
3. Replace the `u.Get("foo_usage_key")` with `r.FooUsageKey`

The problem is some codes do not use the direct string key in the code and do sth like:
```
usageKey := getUsageKey()
u.Get(usageKey)
```
or
```
for usageKey := range usageKeys {
  u.Get(usageKey)
}
```
which makes the analysis harder. However, there are tons of resources not using such complicated logics and we can probably convert them easily.

**Comment by sinabakh** [comment]
# Challenge 1 - d & u Migrations
One of the most challenging steps of migration is how we should handle `d.Get` and `u.Get` method calls. The migration has a few steps:

## Step 1 - Args filling
The first step is to extract all calls and have a 1-o-1 args for each key. So for a resource with the `d.Get("foo").String()` and `u.Get("bar").String()` we will have:
```go
r := &provider.Resource{
  Foo: d.Get("foo").String(),
  Bar: u.Get("bar").String()
}
```
PS: There are surely some difficulties in A) deciding the field name from the key and B) Doing null handling but none of them seems to be impossible to do.

## Step 2 - Replacing the calls with the Args
When moving the logic to the new resource structure, we have no access to `d` and `u` and they are replaced with Args, so we need to replace all occurrences of such method calls with the related Args key.

Before:
```go
if d.Get("foo").String() == "bar" {
  doSth()
}
```

After:
```go
if r.Foo == "bar" {
  doSth()
}
```

## The Challenge - Methods with Complex Args
The challenge we might face is when the method calls arguments are not string literals so we have such calls: `d.Get(someFunc()).String()`, this way there is no easy way that I'm aware of to easily get all possible return values of the function.

## Impact Analysis
I've written a small script that checks whether such a problem exists in a resource or not so we can know how many resources migrations are blocked by it. Here are the results:

AWS:
```
89 of 90 resources can be migrated! The impossible files are: 
 internal/providers/terraform/aws/backup_vault.go
```

Google:
```
40 of 41 resources can be migrated! The impossible files are: 
 internal/providers/terraform/google/container_cluster.go
```

Azure:
```
89 of 90 resources can be migrated! The impossible files are: 
 internal/providers/terraform/azure/synapse_workspace.go
```

So we can theoretically migrate 218 of 221 resources.
@aliscott @tim775 Could you take a look at this comment to see whether I'm missing sth for this step of migration?

**Reply by alikhajeh1** [reply]
This is great! Re null handling, I added [this issue](https://github.com/infracost/infracost/issues/1136) as it seems like we have helper functions but we're not using them consistently so we see occasional bugs. It might be good to consider that when we're migrating.

cc'ing @vdmgolub as he recently worked on [migrating a resource](https://github.com/infracost/infracost/pull/1141) in case he has other ideas.
